### ISSUE TITLE
Tests are failing for configurations that don't use root as database user.

### ISSUE THREAD

**Issue by mdwheele** [opencfp/opencfp#721]
# Steps required to reproduce the problem

1. Use Homestead (that has user `homestead:secret`, not `root:`) or any LAMP stack that doesn't use `root` database user.
2. Clone project and `script/test`

# Expected Result

* I should get similar results to TravisCI

# Actual Result

* I get a ton of DB failures related to database configuration:
```
PDOException: SQLSTATE[HY000] [1045] Access denied for user 'root'@'localhost' (using password: NO) in /home/vagrant/Code/opencfp/ven
dor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOConnection.php:43
```

See this diff where functionality was "broken": https://github.com/opencfp/opencfp/commit/40106c02cef65c094ee470de952284a839bc5fb3#diff-1bd3d4d018fb9e3b5295167d8a53e230

I can resolve this, but I also propose that we make a `testing.yml.dist` similar to other configs so that I can set my own settings locally. I don't use database with root users like our TravisCI config does.

**Comment by mdwheele** [comment]
Thoughts:

* Is there any reason for both `DataBaseInteraction` and `RefreshDatabase` to exist? They're almost doing exactly the same thing, except `RefreshDatabase` is expecting an Illuminate Manager to exist in the container.
* Is there any benefit to maintaining a `dump.sql` file rather than simply dropping all tables and running migrations as usual?

**Comment by mdwheele** [comment]
I'll make a pass at this soon. My current direction will be:

- Get a passing suite by allowing a `root` database user.
- Eliminate `DataBaseInteraction` and substitute `RefreshDatabase` at every usage. Fix broken that results from that.
- Modify `RefreshDatabase` to behave like the following:
  1. Make sure database is migrated exactly once.
  2. Wrap test methods in transactions.
  3. End-state should be a fully-migrated database with only seed data.
- Make `testing` environment database configuration variable.
- Verify tests pass. 
- Fight TravisCI for honor.

**Comment by BackEndTea** [comment]
He difference between the 2 traits is that one will set up the db before every test. The other one sets up the db before the class. This was done since some tests just read from the db to test things.

The dump.sql file is there to speed up the tests. Its faster then migrating up for every test. Since that inckudes a lot of changes and all the aurports. Though i agreee its a bit of a crude option to use an sql file.

On transactions, would we still be able to test for example the delete action in the speaker controller, whih is wrapped in a transaction

**Comment by mdwheele** [comment]
> For transactions: would we still be able to test things like the delete action in the speaker controller, which is wrapped in a transaction

Transaction-ception generally works as long as you [avoid statements that implicitly COMMIT](https://dev.mysql.com/doc/refman/5.7/en/implicit-commit.html). I work on another project that uses transactions to commit events to an `EventStore` and we migrate once at beginning of suite, wrap all test methods in a transaction and tear down the testing database at the end of suite. It works, but admittedly, an event store implementation with MySQL is fairly simplistic (but so should a `DELETE`, in our case).

I'll make sure the suite passes while I'm working on this and afterwards, specifically verify that I'm not getting a false positive on that test.

The one "bad thing" about using transactions is that you don't get to see the state of the database when tests fail (say, if you want to just stop on failure and inspect yourself). But it's not hard to build simple helpers for that.

**Comment by mdwheele** [comment]
Hehe, I've made some pretty good progress with this post Symfony merge. Somewhere since 20 days ago, a change has caused my Docker-based environment to take 7m30s to run the integration suite (up from ~30 seconds). I'm blaming Docker at the moment, but I have been able to reduce runtime to 2m43s. 

I'm going to try the RHEL release of Docker tomorrow and see if there are differences. I'm currently running an edge build for Ubuntu 17.10, soooooo... magic.

If I hold my environment as a constant, the speed improvement is pretty good I think and should have a positive impact on a non-rubbish environment :grin: One side-effect that I didn't account for was how reliant our test-suite was on `AUTO_INCREMENT` values. Things like `User#getOtherTalks` where a specific ID is passed in will fail because tables aren't dropped and reloaded from the side. Transaction rollbacks do not restore `AUTO_INCREMENT` values. Because of the speed improvement so far, I think I'm just going to patch up the tests. If nothing else, it'll reduce the amount of magic numbers.
